design_docs/analysis_task_template.md

```diff
@@ -0,0 +1,741 @@
+        generate = False
+        for element in generateList:
+            if '_' in element:
+                (prefix, suffix) = element.split('_', 1)
+            else:
+                prefix = element
+                suffix = None
+
+            allSuffixes = [self.componentName]
+            if self.categories is not None:
+                allSuffixes = allSuffixes + self.categories
+            noSuffixes = [self.taskName] + allSuffixes
+            if prefix == 'all':
+                if (suffix in allSuffixes) or (suffix is None):
+                    generate = True
+            elif prefix == 'no':
+                if suffix in noSuffixes:
+                    generate = False
+            elif element == self.taskName:
+                generate = True
+
+        return generate  # }}}
+# }}}
+
+# vim: foldmethod=marker ai ts=4 sts=4 et sw=4 ft=python
+```
+
+And here is the suggested template in full:
+
+```python
+"""
+This is an example analysis task to be used as a template for new tasks.
+It should be copied into one of the component folders (`ocean`, `sea_ice`,
+`land_ice`, etc.) and modified as needed.
+
+Don't forget to remove this docstring. (It's not needed.)
+
+Authors
+-------
+Xylar Asay-Davis
+
+Last Modified
+-------------
+03/16/2017
+"""
+
+# import python modules here
+
+# import mpas_analysis module here (those with relative paths starting with
+# dots)
+from ..shared.analysis_task import AnalysisTask
+
+
+class MyTask(AnalysisTask):  # {{{
+    """
+    <Describe the analysis task here.>
+
+    Authors
+    -------
+    <List of authors>
+
+    Last Modified
+    -------------
+    <MM/DD/YYYY>
+    """
+    def __init__(self, config, streamMap=None, variableMap=None,
+                 myArg='myDefaultValue'):  # {{{
+        """
+        Construct the analysis task.
+
+        Parameters
+        ----------
+        config :  instance of MpasAnalysisConfigParser
+            Contains configuration options
+
+        streamMap : dict, optional
+            A dictionary of MPAS-O stream names that map to their mpas_analysis
+            counterparts.
+
+        variableMap : dict, optional
+            A dictionary of MPAS-O variable names that map to their
+            mpas_analysis counterparts.
+
+        myNewArg : str, optional
+            <Describe the arg>
+
+        Authors
+        -------
+        <List of authors>
+
+        Last Modified
+        -------------
+        <MM/DD/YYYY>
+        """
+        # first, call the constructor from the base class (AnalysisTask)
+        AnalysisTask.__init__(config, streamMap, variableMap)
+
+        # next, name the task, the component (ocean, sea_ice, etc.) and the
+        # categories (if any) of the component ('timeSeries', 'climatologyMap'
+        # etc.)
+        self.taskName = 'myTask'
+        self.componentName = 'component'
+        self.categories = ['category1', 'category2']
+        
+        # then, store any additional arguments for use later on
```

Comment:
Also, maybe comments of this type belong in the PR with the code rather than the design doc?  It doesn't really matter if the code in the design doc is the exact code that goes into the PR.  It's just supposed to give you a pretty good idea of what to expect.  So comments here should be about the overall design of the template and comments about individual lines should probably be in #128 .